hashes/src/lib.rs

```diff
@@ -125,101 +119,109 @@ pub mod siphash24;
 
 use core::{convert, fmt, hash};
 
-pub use hmac::{Hmac, HmacEngine};
+#[rustfmt::skip]                // Keep public re-exports separate.
+#[doc(inline)]
+pub use self::sha256t::Tag;
+
+/// Type alias for the [`sha1::Hash`] hash type.
+pub type Sha1 = sha1::Hash;
+
+/// Type alias for the [`sha256::Hash`] hash type.
+pub type Sha256 = sha256::Hash;
+
+/// Type alias for the [`sha384::Hash`] hash type.
+pub type Sha384 = sha384::Hash;
+
+/// Type alias for the [`sha512::Hash`] hash type.
+pub type Sha512 = sha512::Hash;
+
+/// Type alias for the [`sha512_256::Hash`] hash type.
+pub type Sha512_256 = sha512_256::Hash;
+
+/// Type alias for the [`ripemd160::Hash`] hash type.
+pub type Ripemd160 = ripemd160::Hash;
+
+/// Type alias for the [`siphash24::Hash`] hash type.
+pub type Siphash24 = siphash24::Hash;
+
+/// Type alias for the [`hmac::Hash`] hash type.
+pub type Hmac<const N: usize> = hmac::Hash<N>;
+
+/// Type alias for the [`sha256t::Hash`] hash type.
+pub type Sha256t<T> = sha256t::Hash<T>; // FIXME: Perhaps `Sha256T` or `TaggedSha256`?
+
+/// Type alias for the [`hash160::Hash`] hash type.
+pub type Hash160 = hash160::Hash;
+
+/// Type alias for the [`sha256d::Hash`] hash type.
+pub type Sha256d = sha256d::Hash; // FIXME: Perhaps `Sha256D` or maybe `Hash256` (and change the module name)?
 
 /// A hashing engine which bytes can be serialized into.
 pub trait HashEngine: Clone + Default {
-    /// Byte array representing the internal state of the hash engine.
-    type MidState;
+    /// The digest returned by this hash engine.
+    ///
+    /// This is expected to be an array.
+    // trait const types don't work as one would think without nightly.
+    // ref: https://users.rust-lang.org/t/error-e0401-cant-use-generic-parameters-from-outer-function/84512
+    type Digest: Copy
+        + Clone
+        + PartialEq
+        + Eq
+        + PartialOrd
+        + Ord
+        + hash::Hash
+        + convert::AsRef<[u8]>;
 
-    /// Outputs the midstate of the hash engine. This function should not be
-    /// used directly unless you really know what you're doing.
-    fn midstate(&self) -> Self::MidState;
+    /// Byte array representing the internal state of the hash engine.
+    type Midstate;
```

Comment:
Does this need to be generic?

Reply:
It could be an inherent method of sha256 only. IIUC it's the only one that actually uses it. (Or just have it as inherent on multiple if there are more of them.)

Reply:
Seems reasonable, its outside my wheelhouse though, @apoelstra can you weigh in please.